lib/stream.c

```diff
@@ -774,9 +774,12 @@ stream_read_try(struct stream *s, int fd, size_t size)
       s->endp += nbytes;
       return nbytes;
     }
-  /* Error: was it transient (return -2) or fatal (return -1)? */
+  /* If we reach this line and read returned OI retry error
+     connection was closed on remote side. Return 0 to allow client
+     code to handle this situation */
   if (ERRNO_IO_RETRY(errno))
-    return -2;
+    return 0;
```

Comment:
My code returns 0, when read return 0.
Original code returns -2 in this case.
Sasha's code returns 0 in this case, but in case of EINTR, EAGAIN it returns 0 too, which is incorrect